### PR TITLE
Android: Use common CBuffer instead of custom RingBuffer

### PR DESCRIPTION
**Pull request by j-santander**
A couple of changes to the android code:

- When I did my first round of modifications to the already existing android code, I needed some decoupling between receiving samples from the network and copying it to the oboe output buffers. To support that I brought a RingBuffer class without realizing there was already one in the common code (`CBuffer` class in `buffer.h`). Later I realized there were some mistakes in my ring buffer... so now I've taken the opportunity to remove it and use the common CBuffer instead.
- Not sure why we were configuring oboe to use float samples. I believe I recall that at that time there was an ongoing effort to move to float everywhere. That change seems to not have been merged yet, so I've simplified the float to int16_t and int16_t to float code by simply configuring oboe to use int16_t samples.

I've verified the build on my terminal (Google Pixel 3a)

### CODE REVIEW

File: android/sound.cpp [hunk 1]
```diff
@@ -249,17 +249,10 @@ oboe::DataCallbackResult CSound::onAudioInput ( oboe::AudioStream* oboeStream, v
...
     // Copy recording data to internal vector
-    for ( int frmNum = 0; frmNum < numFrames; ++frmNum )
-    {
-        for ( int channelNum = 0; channelNum < oboeStream->getChannelCount(); channelNum++ )
-        {
-            vecsTmpInputAudioSndCrdStereo[frmNum * oboeStream->getChannelCount() + channelNum] =
-                static_cast<int16_t>(floatData[frmNum * oboeStream->getChannelCount() + channelNum] * _MAXSHORT);
-        }
-    }
+    memcpy(vecsTmpInputAudioSndCrdStereo.data(),floatData,sizeof(int16_t)*numFrames * oboeStream->getChannelCount());
```
Comment (ann0see) [inline android/sound.cpp:255]: I think the style is not correct here.

**Review by ann0see — COMMENTED** [review]
Thank you! Didn't check it logic wise, but please have a look at the styling.

File: android/sound.cpp (hunk 1 above)
Comment (j-santander) [reply android/sound.cpp:255]: Thanks, is there a .clang-format I can use? I believe I saw something in the issues/discussions but I don't know if it finally materialized.

Comment (hoffie) [reply android/sound.cpp:255]: Yes, it's here: https://github.com/jamulussoftware/jamulus/pull/1127/files#diff-1026e0038b722990204a42bed8a6f7c0ec2302aa79e3fad1959d62ba968edfa2

It's still not merged and done for existing code yet, but new code can certainly be run through this.

Comment (j-santander) [reply android/sound.cpp]: ok, new commit after applying .clang-format as indicated by @hoffie 
Thanks

File: android/sound.cpp [hunk 2]
```diff
@@ -130,24 +123,15 @@ void CSound::printStreamDetails ( oboe::ManagedStream& stream )
...
-               ", BytesPerFrame: "             << sBytesPerFrame <<
-               ", BytesPerSample: "            << sBytesPerSample <<
-               ", BufferCapacityInFrames: "    << sBufferCapacityInFrames <<
-               ", PerformanceMode: "           << sPerformanceMode <<
-               ", SharingMode: "               << sSharingMode <<
-               ", DeviceID: "                  << sDeviceID <<
-               ", SampleRate: "                << sSampleRate <<
-               ", AudioFormat: "               << sAudioFormat <<
-               ", FramesPerCallback: "         << sFramesPerCallback << "]";
+    qInfo() << "Stream details: [sDirection: " << sDirection << ", FramesPerBurst: " << sFramesPerBurst << ", BufferSizeInFrames: " << sBufferSizeInFrames
```
Comment (pljones) [inline android/sound.cpp:128]: OK, this is a good test for the .clang-format.  The width is just a little too high - this exceeds Github's unified diff frame width.

File: android/sound.cpp [hunk 3]
```diff
@@ -195,17 +179,15 @@ int CSound::Init ( const int iNewPrefMonoBufferSize )
...
-    mOutBuffer.reset ( iOboeBufferSizeStereo * RING_FACTOR );
+    mOutBuffer.Init ( iOboeBufferSizeStereo * RING_FACTOR );
 
     return iOboeBufferSizeMono;
 }
 
 // This is the main callback method for when an audio stream is ready to publish data to an output stream
 // or has received data on an input stream. As per manual much be very careful not to do anything in this back that
 // can cause delays such as sleeping, file processing, allocate memory, etc.
-oboe::DataCallbackResult CSound::onAudioReady ( oboe::AudioStream* oboeStream,
```
Comment (pljones) [inline android/sound.cpp:206]: And I guess this is one of the things we may have to live with going with the .clang-format.

**Comment by pljones** [comment]
@sthenos - as you did the original Android work, is there a chance you could review this, please?

Looks okay to me but I've no Android knowledge.

File: android/sound.cpp [hunk 4]
```diff
@@ -63,7 +63,7 @@ void CSound::closeStream ( oboe::ManagedStream& stream )
         oboe::Result requestStopRes = stream->requestStop();
         oboe::Result result         = stream->close();
 
-        if ( result != oboe::Result::OK )
+        if ( result != oboe::Result::OK || requestStopRes != oboe::Result::OK )
```
Comment (pljones) [inline android/sound.cpp:66]: Would it be worth checking the result of each call:
```
oboe::Result result;
result = stream->requestStop();
if ( oboe::Result::OK != result )
{
    throw CGenErr ( tr ( "Error requesting stream stop: $s", oboe::convertToText ( result ) ) );
}
result = stream->close();
if ( oboe::Result::OK != result )
{
    throw CGenErr ( tr ( "Error closing stream: $s", oboe::convertToText ( result ) ) );
}
```

File: android/sound.cpp [hunk 5]
```diff
@@ -236,17 +236,10 @@ oboe::DataCallbackResult CSound::onAudioInput ( oboe::AudioStream* oboeStream, v
     // We're good to start recording now
     // Take the data from the recording device output buffer and move
     // it to the vector ready to send up to the server
-    float* floatData = static_cast<float*> ( audioData );
+    int16_t* floatData = static_cast<int16_t*> ( audioData );
```
Comment (pljones) [inline android/sound.cpp:239]: Is it floatData in the buffer?

Is it a pointer to integers?

Who knows?

Can you clarify this, please.

Comment (pljones) [reply android/sound.cpp:250]: Why is the _variable name_ `floatData`, if it's an array of `int16_t`?  That's misleading.

Comment (j-santander) [reply android/sound.cpp]: Fixed.

File: android/sound.cpp [hunk 6]
```diff
@@ -45,7 +45,9 @@ void CSound::setupCommonStreamParams ( oboe::AudioStreamBuilder* builder )
 {
     // We request EXCLUSIVE mode since this will give us the lowest possible
     // latency. If EXCLUSIVE mode isn't available the builder will fall back to SHARED mode
-    builder->setFormat ( oboe::AudioFormat::Float )
+
+    // Setting format to be PCM 16 bits (int16_t)
```
Comment (pljones) [inline android/sound.cpp:49]: Why not follow Jamulus elsewhere and use float?  This needs a "...because".

**Comment by pljones** [comment]
OK, one more question raised in the code - but it's probably an example showing clearly someone who knows Android (i.e. not me!) needs to review this.

**Comment by j-santander** [comment]
Sorry, I'm currently away from my desk and I cannot check, but I thought it
was the other way: Android code was using float while the rest of jamulus
used pcm16.

The original (non working) code used float. When I introduced the small
notifications that make it start working I commented about that and I was
told that there was an upcoming change to move everything to float.

When I wrote this PR, I checked and the change to float hadn't happened, so
I removed the extra code converting from float to int and configured oboe
to use pcm16.

When this week I updated the PR I didn't check again if jamulus was still
using pcm16 or was already using floats.

When I get back home I'll review that.

**Comment by pljones** [comment]
Thanks.  It may have been as you say when the original Android branch was taken - it's not that long since Jamulus switched from int to float.

**Comment by j-santander** [comment]
Ok, thanks. I'll review and modify accordingly and resubmit once is ready
and retested.

Cheers

Julian

**Comment by j-santander** [comment]
Oh, well, I'm puzzled.

I see that on `soundbase.h` there's still 
```c++
void ProcessCallback ( CVector<int16_t>& psData )
```

That, I believe, still indicates that (at least the input) is still expected to be int16_t data, not float.

I see that the Issue #544 was closed and the related PR #535 and #627 were merged on a separate branch (no longer active). In PR #660 it is mentioned that a previous attempt (#535) had to be reverted and this PR was closed with

> The branch https://github.com/corrados/jamulus/tree/integrate_float2 is already in the Jamulus Git repo. So there is no need for a Pull Request here. As soon as I continue working on that topic, I'll update the branch and simply merge the code into the Git master branch. I'll close this Pull Request therefore now.

So.... I'd say that the change to float never happened.

**Comment by pljones** [comment]
> So.... I'd say that the change to float never happened.

Thanks for checking (and no wonder I was confused by the sound of it!).

**Comment by ann0see** [comment]
@j-santander @pljones What's missing here?

**Comment by pljones** [comment]
Thanks for pinging me on this -- as it reminds me _again_ that we're not yet float.  At some point that work needs to be completed.  But that's for a different time..!

In terms of review: it needs someone who knows the Android audio subsystem and how to do Android development to comment on it.

**Comment by ann0see** [comment]
> @sthenos - as you did the original Android work, is there a chance you could review this, please?

It would be really great if someone with android knowledge could move this forward.

**Comment by sthenos** [comment]
I'll have a look, when i get a chance.

Cheers
Simon

**Comment by softins** [comment]
> Thanks for pinging me on this -- as it reminds me _again_ that we're not yet float. At some point that work needs to be completed. But that's for a different time..!

The float work that _was_ done was specifically in the mixing (`CServer::MixEncodeTransmitData()`), in order to avoid potential 16-bit overflow when summing channels. Volker's original commit to do this was 9c5a77940f690120a185738d49b8ed4dd6bd3e68, where he starting using an intermediate buffer of `double` instead of summing directly into the `int16` output buffer. Using `double` was probably overkill, and there was a subsequent accepted commit 9171caa417e9e6f6176068e58a8aca9258e398c8 from HPS changing it from `double` to `float`.

Once the final sum of the samples was calculated, it then was converted back to `int16` when writing to the output buffer. 

Of course, the protocol audio data is Opus encoded, so the remaining question would be whether it is possible/better/worse to get the uncompressed audio in and out of Opus in `float` or `int16` format. But as said above, that is for another discussion, not this PR.

**Comment by softins** [comment]
I have squashed this into a single commit from the branch point, and included a necessary fix to `Jamulus.pro`, in my own repo at https://github.com/softins/jamulus/tree/android-update. I retained @j-santander as author of the commit.

I will get it rebased to the current `master` and submit it as a new PR from my branch, to replace this PR.

**Pull request closed without merging**